x/pubsub/keeper/msg_server.go

```diff
@@ -0,0 +1,360 @@
+		return nil, sdkerrors.Wrapf(types.ErrInvalid, "invalid publisher intent: %s", err.Error())
+	}
+
+	publisher, found := k.GetPublisher(ctx, publisherIntent.PublisherDomain)
+	if !found {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrNotFound, "no publisher found with domain: %s", publisherIntent.PublisherDomain)
+	}
+
+	signer := msg.MustGetSigner()
+	if signer.String() != publisher.Address {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrUnauthorized, "registered publisher address must be signer: %s", publisher.Address)
+	}
+
+	_, found = k.GetPublisherIntent(ctx, publisherIntent.SubscriptionId, publisherIntent.PublisherDomain)
+	if found {
+		// TODO(bolten): would this UX be better if it removed the old one for you? deleting a publisher intent will clear out
+		// any subscriber intents attached to it
+		return nil, sdkerrors.Wrapf(types.ErrAlreadyExists, "publisher already has intent for this subscription ID, must be removed first")
+	}
+
+	k.SetPublisherIntent(ctx, *publisherIntent)
+
+	ctx.EventManager().EmitEvents(
+		sdk.Events{
+			sdk.NewEvent(
+				sdk.EventTypeMessage,
+				sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),
+				sdk.NewAttribute(sdk.AttributeKeyAction, types.EventTypeAddPublisherIntent),
+				sdk.NewAttribute(sdk.AttributeKeySender, msg.Signer),
+			),
+			sdk.NewEvent(
+				types.EventTypeAddPublisherIntent,
+				sdk.NewAttribute(types.AttributeKeySubscriptionId, publisherIntent.SubscriptionId),
+				sdk.NewAttribute(types.AttributeKeyPublisherDomain, publisherIntent.PublisherDomain),
+			),
+		},
+	)
+
+	return &types.MsgAddPublisherIntentResponse{}, nil
+}
+
+func (k Keeper) AddSubscriberIntent(c context.Context, msg *types.MsgAddSubscriberIntent) (*types.MsgAddSubscriberIntentResponse, error) {
+	ctx := sdk.UnwrapSDKContext(c)
+	subscriberIntent := msg.SubscriberIntent
+
+	if err := subscriberIntent.ValidateBasic(); err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrInvalid, "invalid subscriber intent: %s", err.Error())
+	}
+
+	subscriptionId := subscriberIntent.SubscriptionId
+	subscriberAddress := subscriberIntent.SubscriberAddress
+	publisherDomain := subscriberIntent.PublisherDomain
+
+	// ValidateBasic will confirm this is already correctly formatted
+	subscriberAccAddress, _ := sdk.AccAddressFromBech32(subscriberIntent.SubscriberAddress)
+	subscriber, found := k.GetSubscriber(ctx, subscriberAccAddress)
+	if !found {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrNotFound, "no subscriber found with address: %s", subscriberAddress)
+	}
+
+	signer := msg.MustGetSigner()
+	signerAddress := signer.String()
+	if signerAddress != subscriber.Address {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrUnauthorized, "registered subscriber address must be signer: %s", subscriber.Address)
+	}
+
+	_, found = k.GetSubscriberIntent(ctx, subscriberIntent.SubscriptionId, subscriberAccAddress)
+	if found {
+		// TODO(bolten): would this UX be better if it removed the old one for you?
```

Comment:
I mean, maybe -- I feel like `AddSubscriberIntent` is effectively writing that subscriber's publisher choice for a given subscription ID, and writing it again changes it rather than manually removing then adding it. If the language is confusing, maybe `Update` makes more sense, even if it didn't originally exist? I dunno, that's why I'm asking the question.

Reply:
To me, Add has a connotation of there potentially being more than one. Update seems more specific

Reply:
For simplicity's sake I think I will leave add/remove as is without creating some kind of "update" concept, at least for now.